sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift

```diff
@@ -29,23 +29,102 @@ import Foundation
 
 /// A Storage shared access signature credential object.
 public struct StorageSASCredential: AzureCredential {
-    internal let blobEndpoint: String?
-    internal let queueEndpoint: String?
-    internal let fileEndpoint: String?
-    internal let tableEndpoint: String?
-    internal let sasToken: String?
-    internal let error: Error?
+    internal let blobSasUriProvider: (() -> String)?
+    internal let connectionStringProvider: (() -> String)?
+    internal var blobEndpoint: String?
+    internal var queueEndpoint: String?
+    internal var fileEndpoint: String?
+    internal var tableEndpoint: String?
+    internal var error: Error?
+    internal var sasToken: String? {
+        didSet { expirationDate = StorageSASCredential.expiration(forToken: sasToken) }
+    }
+
+
+    // Treat tokens that expire in <= 2 minutes as expired
```

Comment:
What's the rationale for this? What is the typical expiration?

Reply:
So, I was under the impression that the server would provide a meaningful error response if the token was expired or otherwise invalid, but it seems that's not (always?) the case. In at least some cases, an expired token simply returns a 404 as if the entity did not exist at all. So the logic for determining whether the token is expired or not is (unfortunately) entirely based on the expiration date stamped into the token itself.

In order to have a greater chance of success and to prevent the customer from getting an unexpected 404 or other error response, we optimistically refresh the token just a little bit before it expires. I'm trying to recall where this value came from, but it was discussed last week and it was suggested that other SDKs used a value of two minutes in order to be as small as possible while mitigating (small amounts of) clock skew between the client and the server.

I would definitely want to get some feedback from other language SDK developers and/or the storage service developers on this.